presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitSource.java

```diff
@@ -54,45 +67,190 @@
+            Executor executor,
+            CounterStat highMemorySplitSourceCounter)
+    {
+        AtomicReference<State> stateReference = new AtomicReference<>(State.initial());
+        return new HiveSplitSource(
+                connectorId,
+                queryId,
+                databaseName,
+                tableName,
+                compactEffectivePredicate,
+                new PerBucket()
+                {
+                    private final AsyncQueue<Optional<InternalHiveSplit>> queue = new AsyncQueue<>(maxOutstandingSplits, executor);
+
+                    public CompletableFuture<?> offer(OptionalInt bucketNumber, Optional<InternalHiveSplit> connectorSplit)
+                    {
+                        // bucketNumber can be non-empty because BackgroundHiveSplitLoader does not have knowledge of execution plan
+                        return queue.offer(connectorSplit);
+                    }
+
+                    @Override
+                    public CompletableFuture<List<Optional<InternalHiveSplit>>> getBatchAsync(OptionalInt bucketNumber, int maxSize)
+                    {
+                        checkArgument(!bucketNumber.isPresent());
+                        return queue.getBatchAsync(maxSize);
+                    }
+
+                    @Override
+                    public void offerAndFinish(Optional<InternalHiveSplit> connectorSplit)
+                    {
+                        queue.offerAndFinish(connectorSplit);
+                    }
+
+                    @Override
+                    public void finish()
+                    {
+                        queue.finish();
+                    }
+
+                    @Override
+                    public int getOutstandingSplitCount()
+                    {
+                        return queue.size();
+                    }
+                },
+                maxOutstandingSplitsSize,
+                splitLoader,
+                stateReference,
+                highMemorySplitSourceCounter);
+    }
+
+    public static HiveSplitSource bucketed(
+            String connectorId,
+            String queryId,
+            String databaseName,
+            String tableName,
+            TupleDomain<? extends ColumnHandle> compactEffectivePredicate,
+            int estimatedOutstandingSplitsPerBucket,
+            DataSize maxOutstandingSplitsSize,
+            HiveSplitLoader splitLoader,
+            Executor executor,
+            CounterStat highMemorySplitSourceCounter)
+    {
+        AtomicReference<State> stateReference = new AtomicReference<>(State.initial());
+        return new HiveSplitSource(
+                connectorId,
+                queryId,
+                databaseName,
+                tableName,
+                compactEffectivePredicate,
+                new PerBucket()
+                {
+                    private final Map<OptionalInt, AsyncQueue<Optional<InternalHiveSplit>>> queues = new ConcurrentHashMap<>();
```

Comment:
Looks like the OptionalInt is always present so why not make this map key Integer